File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
    if (s.isEmpty()) {
        return 0;
    }

    int maxans = 0;  // Start from 0 instead of Integer.MIN_VALUE

    for (int i = 0; i < s.length(); i++) {
        Set<Character> se = new HashSet<>();
        for (int j = i; j < s.length(); j++) {
            if (se.contains(s.charAt(j))) {
                break;  // Break if a duplicate is found
            }
            se.add(s.charAt(j));
            maxans = Math.max(maxans, j - i + 1); // Update maxans for every valid substring
        }
    }

    return maxans;

    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        hash = {}
        left = 0
        max_len = 0
        for right in range(n):
            if s[right] in hash:
                left = max(left, hash[s[right]]+1)
            hash[s[right]] = right

            max_len = max(max_len, right-left+1)

        return max_len